File: dart/dartfind/lib/src/finder.dart
import 'dart:async';
import 'dart:io';

import 'package:dartfind/dartfind.dart';
import 'package:path/path.dart' as path;

const String startPathNotDefined = 'Startpath not defined';
const String startPathNotFound = 'Startpath not found';
const String startPathNotMatchFindSettings =
    'Startpath does not match find settings';
const String invalidRangeDepth = 'Invalid range for mindepth and maxdepth';
const String invalidRangeLastMod =
    'Invalid range for minlastmod and maxlastmod';
const String invalidRangeSize = 'Invalid range for minsize and maxsize';

class Finder {
  final FindSettings settings;
  final FileTypes _fileTypes = FileTypes();
  // whether to allow invalid/malformed characters in encoding/decoding
  // setting to false means find will end once an attempt is made to read
  // a file in an incompatible encoding
  final bool allowInvalid = true;
  Finder(this.settings) {
    _validateSettings();
  }

  void _validateSettings() {
    if (settings.paths.isEmpty) {
      throw FindException(startPathNotDefined);
    }
    for (var path in settings.paths) {
      if (path == null || path.trim().isEmpty) {
        throw FindException(startPathNotFound);
      }
      var p = path;
      var pathType =
          FileSystemEntity.typeSync(p, followLinks: settings.followSymlinks);
      if (pathType == FileSystemEntityType.notFound) {
        p = FileUtil.expandPath(p);
        pathType =
            FileSystemEntity.typeSync(p, followLinks: settings.followSymlinks);
        if (pathType == FileSystemEntityType.notFound) {
          throw FindException(startPathNotFound);
        }
      }
      if (pathType == FileSystemEntityType.link) {
        // this means followSymlinks is false,
        throw FindException(startPathNotMatchFindSettings);
      } else if (pathType == FileSystemEntityType.directory) {
        if (!isTraversableDir(Directory(p))) {
          throw FindException(startPathNotMatchFindSettings);
        }
      } else if (pathType == FileSystemEntityType.file) {
        if (!isMatchingFile(File(p))) {
          throw FindException(startPathNotMatchFindSettings);
        }
      } else {
        // TODO: start path is unknown/invalid type
        throw FindException(startPathNotMatchFindSettings);
      }
    }
    if (settings.maxDepth > -1 &&
        settings.minDepth > -1 &&
        (settings.maxDepth < settings.minDepth)) {
      throw FindException(invalidRangeDepth);
    }
    if (settings.maxLastMod != null &&
        settings.minLastMod != null &&
        (settings.maxLastMod!.millisecondsSinceEpoch <
            settings.minLastMod!.millisecondsSinceEpoch)) {
      throw FindException(invalidRangeLastMod);
    }
    if (settings.maxSize > 0 &&
        settings.minSize > 0 &&
        (settings.maxSize < settings.minSize)) {
      throw FindException(invalidRangeSize);
    }
  }

  bool _matchesAnyPattern(String s, Set<Pattern> patterns) {
    return patterns.any((p) => p.allMatches(s).isNotEmpty);
  }

  bool _anyMatchesAnyPattern(Iterable<String> elems, Set<Pattern> patterns) {
    return elems.any((elem) => _matchesAnyPattern(elem, patterns));
  }

  bool _emptyOrMatchesAnyPattern(String s, Set<Pattern> patterns) {
    return patterns.isEmpty || _matchesAnyPattern(s, patterns);
  }

  bool _emptyOrNotMatchesAnyPattern(String s, Set<Pattern> patterns) {
    return patterns.isEmpty || !_matchesAnyPattern(s, patterns);
  }

  bool _emptyOrAnyMatchesAnyPattern(
      Iterable<String> elems, Set<Pattern> patterns) {
    return patterns.isEmpty || _anyMatchesAnyPattern(elems, patterns);
  }

  bool _emptyOrNotAnyMatchesAnyPattern(
      Iterable<String> elems, Set<Pattern> patterns) {
    return patterns.isEmpty || !_anyMatchesAnyPattern(elems, patterns);
  }

  bool _emptyOrMatchesAnyString(String s, Set<String> strings) {
    return strings.isEmpty || strings.contains(s);
  }

  bool _emptyOrNotMatchesAnyString(String s, Set<String> strings) {
    return strings.isEmpty || !strings.contains(s);
  }

  bool _emptyOrMatchesAnyFileType(FileType s, Set<FileType> fileTypes) {
    return fileTypes.isEmpty || fileTypes.contains(s);
  }

  bool _emptyOrNotMatchesAnyFileType(FileType s, Set<FileType> fileTypes) {
    return fileTypes.isEmpty || !fileTypes.contains(s);
  }

  bool isMatchingDirByHidden(Directory dir) {
    return settings.includeHidden || !FileUtil.isHiddenPath(dir.path);
  }

  bool isMatchingDirByInPatterns(Directory dir) {
    var elems = FileUtil.pathElems(dir.path);
    return _emptyOrAnyMatchesAnyPattern(elems, settings.inDirPatterns);
  }

  bool isMatchingDirByOutPatterns(Directory dir) {
    var elems = FileUtil.pathElems(dir.path);
    return _emptyOrNotAnyMatchesAnyPattern(elems, settings.outDirPatterns);
  }

  bool isTraversableDir(Directory dir) {
    return isMatchingDirByHidden(dir) && isMatchingDirByOutPatterns(dir);
  }

  bool isMatchingDir(Directory dir) {
    return isMatchingDirByHidden(dir) &&
        isMatchingDirByInPatterns(dir) &&
        isMatchingDirByOutPatterns(dir);
  }

  bool isNullOrMatchingDir(Directory? dir) {
    return dir == null || isMatchingDir(dir);
  }

  bool isMatchingFileNameByHidden(String fileName) {
    return settings.includeHidden || !FileUtil.isHiddenName(fileName);
  }

  bool isMatchingArchiveExtension(String ext) {
    return _emptyOrMatchesAnyString(ext, settings.inArchiveExtensions) &&
        _emptyOrNotMatchesAnyString(ext, settings.outArchiveExtensions);
  }

  bool isMatchingArchiveExtensionForFile(File file) {
    if (settings.inArchiveExtensions.isNotEmpty ||
        settings.outArchiveExtensions.isNotEmpty) {
      var fileName = path.basename(file.path);
      var ext = FileUtil.extension(fileName);
      return isMatchingArchiveExtension(ext);
    }
    return true;
  }

  bool isMatchingArchiveFileName(String fileName) {
    return _emptyOrMatchesAnyPattern(
            fileName, settings.inArchiveFilePatterns) &&
        _emptyOrNotMatchesAnyPattern(fileName, settings.outArchiveFilePatterns);
  }

  bool isMatchingArchiveFileNameForFile(File file) {
    if (settings.inArchiveFilePatterns.isNotEmpty ||
        settings.outArchiveFilePatterns.isNotEmpty) {
      var fileName = path.basename(file.path);
      return isMatchingArchiveFileName(fileName);
    }
    return true;
  }

  bool isMatchingArchiveFile(File file) {
    return isNullOrMatchingDir(file.parent) &&
        isMatchingArchiveExtensionForFile(file) &&
        isMatchingArchiveFileNameForFile(file);
  }

  bool isMatchingArchiveFileResult(FileResult fr) {
    return isMatchingArchiveFile(fr.file);
  }

  bool isMatchingExtension(String ext) {
    return _emptyOrMatchesAnyString(ext, settings.inExtensions) &&
        _emptyOrNotMatchesAnyString(ext, settings.outExtensions);
  }

  bool isMatchingExtensionForFile(File file) {
    if (settings.inExtensions.isNotEmpty || settings.outExtensions.isNotEmpty) {
      var fileName = path.basename(file.path);
      var ext = FileUtil.extension(fileName);
      return isMatchingExtension(ext);
    }
    return true;
  }

  bool isMatchingFileName(String fileName) {
    return _emptyOrMatchesAnyPattern(fileName, settings.inFilePatterns) &&
        _emptyOrNotMatchesAnyPattern(fileName, settings.outFilePatterns);
  }

  bool isMatchingFileNameForFile(File file) {
    if (settings.inFilePatterns.isNotEmpty ||
        settings.outFilePatterns.isNotEmpty) {
      var fileName = path.basename(file.path);
      return isMatchingFileName(fileName);
    }
    return true;
  }

  bool isMatchingFile(File file) {
    return isNullOrMatchingDir(file.parent) &&
        isMatchingExtensionForFile(file) &&
        isMatchingFileNameForFile(file);
  }

  bool isMatchingFileType(FileType fileType) {
    return _emptyOrMatchesAnyFileType(fileType, settings.inFileTypes) &&
        _emptyOrNotMatchesAnyFileType(fileType, settings.outFileTypes);
  }

  bool isMatchingFileSize(int fileSize) {
    return ((settings.maxSize == 0 || fileSize <= settings.maxSize) &&
        (settings.minSize == 0 || fileSize >= settings.minSize));
  }

  bool isMatchingLastMod(DateTime? lastMod) {
    return ((settings.maxLastMod == null ||
            lastMod!.compareTo(settings.maxLastMod!) <= 0) &&
        (settings.minLastMod == null ||
            lastMod!.compareTo(settings.minLastMod!) >= 0));
  }

  bool isMatchingFileResult(FileResult fr) {
    return isMatchingFile(fr.file) &&
        isMatchingFileType(fr.fileType) &&
        isMatchingFileSize(fr.fileSize) &&
        isMatchingLastMod(fr.lastMod);
  }

  (int, DateTime?) getFileSizeAndLastMod(File f) {
    int fileSize = 0;
    DateTime? lastMod;
    if (settings.needSize() || settings.needLastMod()) {
      FileStat stat = FileStat.statSync(f.path);
      if (settings.needSize()) {
        fileSize = stat.size;
      }
      if (settings.needLastMod()) {
        lastMod = stat.modified;
      }
    }
    return (fileSize, lastMod);
  }

  Future<FileResult?> filterArchiveFileToFileResult(File f) async {
    if (!settings.includeArchives && !settings.archivesOnly) {
      return Future.value(null);
    }

    if (!isMatchingArchiveFile(f)) {
      return Future.value(null);
    }

    int fileSize = 0;
    DateTime? lastMod;
    var fileResult = FileResult(f, FileType.archive, fileSize, lastMod);
    return Future.value(fileResult);
  }

  Future<FileResult?> filterRegularFileToFileResult(
      File f, FileType fileType) async {
    if (settings.archivesOnly) {
      return Future.value(null);
    }

    if (!isMatchingFile(f) || !isMatchingFileType(fileType)) {
      return Future.value(null);
    }

    var (fileSize, lastMod) = getFileSizeAndLastMod(f);

    if (!isMatchingFileSize(fileSize) || !isMatchingLastMod(lastMod)) {
      return Future.value(null);
    }

    var fileResult = FileResult(f, fileType, fileSize, lastMod);
    return Future.value(fileResult);
  }

  Future<FileResult?> filterToFileResult(File f) async {
    if (!isNullOrMatchingDir(f.parent)) {
      // return Future.value(null);
      return null;
    }

    var fileName = path.basename(f.path);
    if (!isMatchingFileNameByHidden(fileName)) {
      // return Future.value(null);
      return null;
    }

    var fileType = await _fileTypes.getFileType(fileName);
    if (fileType == FileType.archive) {
      return filterArchiveFileToFileResult(f);
    }
    return filterRegularFileToFileResult(f, fileType);
  }

  Future<List<FileResult>> filterToFileResults(List<File> filePaths) {
    var fileResultsFutures = filePaths.map((f) => filterToFileResult(f));
    return Future.wait(fileResultsFutures).then((fileResults) {
      return fileResults.where((fr) => fr != null).toList().cast<FileResult>();
    });
  }

  Future<List<FileResult>> _recGetFileResultsForDir(
      Directory dir, int minDepth, int maxDepth, int currentDepth) async {
    var fileResults = <FileResult>[];
    var recurse = true;
    if (currentDepth == maxDepth) {
      recurse = false;
    } else if (maxDepth > -1 && currentDepth > maxDepth) {
      return [];
    }

    // Get dirs and files under filePath
    var pathDirs = <Directory>[];
    var pathFiles = <File>[];
    List<FileSystemEntity> entries = dir
        .listSync(recursive: false, followLinks: settings.followSymlinks)
        .toList();
    for (var entry in entries) {
      var linkIsDir = false;
      var linkIsFile = false;
      // TODO: I think this is supposed to mean followSymlinks == false?
      if (entry is Link) {
        var resolvedPath = entry.resolveSymbolicLinksSync();
        if (resolvedPath is Directory) {
          linkIsDir = true;
        } else if (resolvedPath is File) {
          linkIsFile = true;
        }
      }
      if ((entry is Directory || linkIsDir) &&
          recurse &&
          isMatchingDirByHidden(entry as Directory) &&
          isMatchingDirByOutPatterns(entry)) {
        pathDirs.add(entry);
      } else if ((entry is File || linkIsFile) &&
          (minDepth < 0 || currentDepth >= minDepth)) {
        pathFiles.add(entry as File);
      }
    }

    // Filter files
    var pathResults = await filterToFileResults(pathFiles);
    fileResults.addAll(pathResults);

    // Recurse into dirs
    for (var pathDir in pathDirs) {
      var subFileResults = await _recGetFileResultsForDir(
          pathDir, minDepth, maxDepth, currentDepth + 1);
      fileResults.addAll(subFileResults);
    }
    return fileResults;
  }

  Future<List<FileResult>> _getFileResultsForPath(String aPath) async {
    var fileSysType = await FileSystemEntity.type(aPath,
        followLinks: settings.followSymlinks);
    if (fileSysType == FileSystemEntityType.notFound) {
      aPath = FileUtil.expandPath(aPath);
      fileSysType = await FileSystemEntity.type(aPath,
          followLinks: settings.followSymlinks);
      if (fileSysType == FileSystemEntityType.notFound) {
        throw FindException(startPathNotFound);
      }
    }

    if (fileSysType == FileSystemEntityType.link) {
      // this means followSymlinks is false,
      throw FindException(startPathNotMatchFindSettings);
    }

    if (fileSysType == FileSystemEntityType.directory) {
      // if max_depth is zero, we can skip since a directory cannot be a result
      if (settings.maxDepth == 0) {
        return [];
      }

      var dir = Directory(aPath);
      if (isTraversableDir(dir)) {
        var maxDepth = settings.maxDepth;
        if (!settings.recursive) {
          maxDepth = 1;
        }
        return _recGetFileResultsForDir(dir, settings.minDepth, maxDepth, 1);
      } else {
        throw FindException(startPathNotMatchFindSettings);
      }
    } else if (fileSysType == FileSystemEntityType.file) {
      // if min_depth > zero, we can skip since the file is at depth zero
      if (settings.minDepth > 0) {
        return [];
      }

      var startFile = File(aPath);
      var fileResult = await filterToFileResult(startFile);

      if (fileResult != null) {
        return [fileResult];
      } else {
        throw FindException(startPathNotMatchFindSettings);
      }
    } else {
      throw FindException(startPathNotMatchFindSettings);
    }
  }

  Future<List<FileResult>> _findFiles() async {
    var fileResults = <FileResult>[];
    var pathsFileResultsFutures =
        settings.paths.map((p) => _getFileResultsForPath(p!));
    await Future.wait(pathsFileResultsFutures).then((pathsFileResults) {
      for (var pathFileResults in pathsFileResults) {
        fileResults.addAll(pathFileResults);
      }
    });
    if (fileResults.length > 1) {
      var fileResultSorter = FileResultSorter(settings);
      fileResultSorter.sort(fileResults);
    }
    return fileResults;
  }

  Future<List<FileResult>> find() async {
    return Future.wait([_fileTypes.ready]).then((res) {
      return _findFiles();
    });
  }

  List<Directory> getMatchingDirs(List<FileResult> fileResults) {
    Map<String, Directory> pathDirMap = {};
    for (var fr in fileResults) {
      var dir = fr.file.parent;
      pathDirMap[dir.path] = dir;
    }
    var paths = pathDirMap.keys.toList();
    paths.sort();
    List<Directory> dirs = [];
    for (var p in paths) {
      var dir = pathDirMap[p];
      dirs.add(dir!);
    }
    return dirs;
  }

  void printMatchingDirs(
      List<FileResult> fileResults, FileResultFormatter formatter) {
    var dirs = getMatchingDirs(fileResults);
    if (dirs.isNotEmpty) {
      logMsg('\nMatching directories (${dirs.length}):');
      for (var d in dirs) {
        logMsg(formatter.formatDirectory(d));
      }
    } else {
      logMsg('\nMatching directories: 0');
    }
  }

  void printMatchingFiles(
      List<FileResult> fileResults, FileResultFormatter formatter) {
    if (fileResults.isNotEmpty) {
      logMsg('\nMatching files (${fileResults.length}):');
      for (var fr in fileResults) {
        logMsg(formatter.formatFileResult(fr));
      }
    } else {
      logMsg('\nMatching files: 0');
    }
  }
}

File: elixir/exfind/lib/finder.ex
defmodule ExFind.Finder do
  @moduledoc """
  Documentation for `ExFind.Finder`.
  """

  alias ExFind.FileResult
  alias ExFind.FileResultFormatter
  alias ExFind.FileResultSorter
  alias ExFind.FileTypes
  alias ExFind.FileUtil
  alias ExFind.FindError
  alias ExFind.FindSettings
  alias ExFind.Logging
  alias ExFind.StringUtil

  defstruct [:file_types, :settings]

  # def new(args), do: __struct__(args)
  def new(settings) do
    file_types = ExFind.FileTypes.new()
    __struct__([file_types: file_types, settings: settings])
  end

  def matches_any_string?(string, strings) do
    Enum.any?(strings, fn s -> s == string end)
  end

  def empty_or_matches_any_string?(string, strings) do
    Enum.empty?(strings) or matches_any_string?(string, strings)
  end

  def empty_or_not_matches_any_string?(string, strings) do
    Enum.empty?(strings) or not matches_any_string?(string, strings)
  end

  def matches_any_pattern?(string, patterns) do
    Enum.any?(patterns, fn p -> Regex.match?(p, string) end)
  end

  def empty_or_matches_any_pattern?(string, patterns) do
    Enum.empty?(patterns) or matches_any_pattern?(string, patterns)
  end

  def empty_or_not_matches_any_pattern?(string, patterns) do
    Enum.empty?(patterns) or not matches_any_pattern?(string, patterns)
  end

  def any_matches_any_pattern?(strings, patterns) do
    Enum.any?(strings, fn s -> matches_any_pattern?(s, patterns) end)
  end

  def empty_or_any_matches_any_pattern?(strings, patterns) do
    Enum.empty?(patterns) or any_matches_any_pattern?(strings, patterns)
  end

  def empty_or_not_any_matches_any_pattern?(strings, patterns) do
    Enum.empty?(patterns) or not any_matches_any_pattern?(strings, patterns)
  end

  def matching_dir_by_hidden?(finder, dir) do
    # IO.puts("matching_dir_by_hidden?(#{dir})")
    (finder.settings.include_hidden or not FileUtil.hidden_path?(dir))
  end

  def matching_dir_by_in_patterns?(finder, dir) do
    # IO.puts("matching_dir_by_in_patterns?(#{dir})")
     empty_or_any_matches_any_pattern?(Path.split(dir), finder.settings.in_dir_patterns)
  end

  def matching_dir_by_out_patterns?(finder, dir) do
    # IO.puts("matching_dir_by_out_patterns?(#{dir})")
    empty_or_not_any_matches_any_pattern?(Path.split(dir), finder.settings.out_dir_patterns)
  end

  defp traversable_dir?(finder, dir) do
    matching_dir_by_hidden?(finder, dir) and matching_dir_by_out_patterns?(finder, dir)
  end

  def matching_dir?(finder, dir) do
    # IO.puts("matching_dir?(#{dir})")
    matching_dir_by_hidden?(finder, dir)
    and matching_dir_by_in_patterns?(finder, dir)
    and matching_dir_by_out_patterns?(finder, dir)
  end

  def null_or_matching_dir?(finder, dir) do
    # IO.puts("null_or_matching_dir?(#{dir})")
    dir == nil or String.trim(dir) == "" or matching_dir?(finder, dir)
  end

  def matching_file_name_by_hidden?(finder, file_name) do
    # IO.puts("matching_file_name_by_hidden?(#{file_name})")
    (finder.settings.include_hidden or not FileUtil.hidden_name?(file_name))
  end

  def matching_archive_extension?(finder, ext) do
    # IO.puts("matching_extension?(#{ext})")
    empty_or_matches_any_string?(ext, finder.settings.in_archive_extensions)
    and empty_or_not_matches_any_string?(ext, finder.settings.out_archive_extensions)
  end

  def has_matching_archive_extension?(finder, file_name) do
    # IO.puts("has_matching_archive_extension?(#{file_name})")
    if not Enum.empty?(finder.settings.in_archive_extensions) or not Enum.empty?(finder.settings.out_archive_extensions) do
      ext = FileUtil.get_extension(file_name)
      matching_archive_extension?(finder, ext)
    else
      true
    end
  end

  def matching_extension?(finder, ext) do
    # IO.puts("matching_extension?(#{ext})")
    empty_or_matches_any_string?(ext, finder.settings.in_extensions)
    and empty_or_not_matches_any_string?(ext, finder.settings.out_extensions)
  end

  def has_matching_extension?(finder, file_name) do
    # IO.puts("has_matching_extension?(#{file_name})")
    if not Enum.empty?(finder.settings.in_extensions) or not Enum.empty?(finder.settings.out_extensions) do
      ext = FileUtil.get_extension(file_name)
      matching_extension?(finder, ext)
    else
      true
    end
  end

  def matching_archive_file_name?(finder, file_name) do
    # IO.puts("matching_file_name?(#{file_name})")
    empty_or_matches_any_pattern?(file_name, finder.settings.in_archive_file_patterns)
    and empty_or_not_matches_any_pattern?(file_name, finder.settings.out_archive_file_patterns)
  end

  def matching_archive_file_path?(finder, file_path) do
    # IO.puts("matching_archive_file_path?(#{file_path})")
    file_name = Path.basename(file_path)
    has_matching_archive_extension?(finder, file_name) and matching_archive_file_name?(finder, file_name)
  end

  # TODO: change this to matching_archive_file_result
  def matching_archive_file_path?(finder, file_path, _file_size, _last_mod) do
    # IO.puts("matching_archive_file_path?(#{file_path})")
    file_name = Path.basename(file_path)
    (finder.settings.include_hidden or not FileUtil.hidden_name?(Path.basename(file_name)))
    and has_matching_archive_extension?(finder, file_name)
    and matching_archive_file_name?(finder, file_name)
  end

  def matching_file_name?(finder, file_name) do
    # IO.puts("matching_file_name?(#{file_name})")
    empty_or_matches_any_pattern?(file_name, finder.settings.in_file_patterns)
    and empty_or_not_matches_any_pattern?(file_name, finder.settings.out_file_patterns)
  end

  def matching_file_path?(finder, file_path) do
    # IO.puts("matching_file_path?(#{file_path})")
    file_name = Path.basename(file_path)
    has_matching_extension?(finder, file_name) and matching_file_name?(finder, file_name)
  end

  def matching_file_type?(finder, file_type) do
    # IO.puts("matching_file_type?(#{file_type})")
    (Enum.empty?(finder.settings.in_file_types) or Enum.any?(finder.settings.in_file_types, fn t -> t == file_type end))
    and (Enum.empty?(finder.settings.out_file_types) or !Enum.any?(finder.settings.out_file_types, fn t -> t == file_type end))
  end

  def matching_file_size?(finder, file_size) do
    # IO.puts("matching_file_size?(#{file_size})")
    (finder.settings.min_size == 0 or file_size >= finder.settings.min_size)
    and (finder.settings.max_size == 0 or file_size <= finder.settings.max_size)
  end

  def matching_last_mod?(finder, last_mod) do
    # IO.puts("matching_last_mod?(#{inspect(last_mod)})")
    (finder.settings.min_last_mod == nil
     or last_mod >= DateTime.to_unix(finder.settings.min_last_mod))
    and (finder.settings.max_last_mod == nil
         or last_mod <= DateTime.to_unix(finder.settings.max_last_mod))
  end

  # TODO: change this to matching_file_result
  def matching_file_path?(finder, file_path, file_type, file_size, last_mod) do
    # IO.puts("matching_file_path?(#{file_path})")
    file_name = Path.basename(file_path)
    (finder.settings.include_hidden or not FileUtil.hidden_name?(Path.basename(file_name)))
    and has_matching_extension?(finder, file_name)
    and matching_file_name?(finder, file_name)
    and matching_file_type?(finder, file_type)
    and matching_file_size?(finder, file_size)
    and matching_last_mod?(finder, last_mod)
  end

  def filter_archive_file_path_to_file_results(finder, file_path) do
    # IO.puts("filter_archive_file_path_to_file_results(#{file_path})")
    if not finder.settings.include_archives and not finder.settings.archives_only do
      []
    else
      if not matching_archive_file_path?(finder, file_path) do
        []
      else
        [FileResult.new(Path.dirname(file_path), Path.basename(file_path), :archive, 0, 0)]
      end
    end
  end

  def filter_regular_file_path_to_file_results(finder, file_path, file_type) do
    # IO.puts("filter_regular_file_path_to_file_results(#{file_path})")
    if finder.settings.archives_only do
      []
    else
      if not matching_file_path?(finder, file_path) or not matching_file_type?(finder, file_type) do
        []
      else
        file_stat =
          if FindSettings.need_size?(finder.settings) or FindSettings.need_last_mod?(finder.settings) do
            File.stat!(file_path, [time: :posix])
          else
            nil
          end
        {file_size, last_mod} =
          if file_stat != nil do
            {file_stat.size, file_stat.mtime}
          else
            {0, 0}
          end
        if not matching_file_size?(finder, file_size) or not matching_last_mod?(finder, last_mod) do
          []
        else
          [FileResult.new(Path.dirname(file_path), Path.basename(file_path), file_type, file_size, last_mod)]
        end
      end
    end
  end

  def filter_to_file_results(finder, file_path) do
    # IO.puts("filter_to_file_results(#{file_path})")
    if not matching_dir?(finder, Path.dirname(file_path)) or not matching_file_name_by_hidden?(finder, Path.basename(file_path)) do
      []
    else
      case FileTypes.get_file_type_for_file_name(finder.file_types, Path.basename(file_path)) do
        :archive -> filter_archive_file_path_to_file_results(finder, file_path)
        file_type -> filter_regular_file_path_to_file_results(finder, file_path, file_type)
      end
    end
  end

  def rec_find_path(finder, path, min_depth, max_depth, current_depth) do
    # IO.puts("rec_find_path(#{path})")
    if max_depth > -1 and current_depth > max_depth do
      []
    else
      {dirs, files} = File.ls!(path)
                      |> Enum.map(fn f -> Path.join(path, f) end)
                      |> Enum.filter(fn f -> File.regular?(f) or File.dir?(f) end)
                      |> Enum.split_with(fn f -> File.dir?(f) end)
      recurse = max_depth == -1 or current_depth < max_depth
      dirs = if recurse, do: dirs, else: []
      dirs = Enum.filter(dirs, fn d -> finder.settings.follow_symlinks or not FileUtil.symlink?(d) end)
             |> Enum.filter(fn d -> traversable_dir?(finder, d) end)

      files = Enum.filter(files, fn f -> finder.settings.follow_symlinks or not FileUtil.symlink?(f) end)

      file_results = if min_depth < 0 or current_depth >= min_depth do
        Enum.map(files, fn f -> filter_to_file_results(finder, f) end) |> List.flatten()
      else
        []
      end
      dir_results = dirs
                    |> Enum.map(fn d -> rec_find_path(finder, d, min_depth, max_depth, current_depth + 1) end)
                    |> List.flatten()
      file_results ++ dir_results
    end
  end

  def find_path!(finder, path) do
    # IO.puts("find_path!(#{path})")
    p = if File.exists?(path) do
      path
    else
      FileUtil.expand_path(path)
    end

    cond do
      not File.exists?(p) -> raise FindError, message: FindError.startpath_not_match_find_settings
      File.dir?(p) ->
        if not finder.settings.follow_symlinks and FileUtil.symlink?(p) do
          raise FindError, message: FindError.startpath_not_match_find_settings
        end
        if finder.settings.max_depth == 0 do
          []
        else
          if traversable_dir?(finder, p) do
            max_depth = if finder.settings.recursive, do: finder.settings.max_depth, else: 1
            rec_find_path(finder, p, finder.settings.min_depth, max_depth, 1)
          else
            raise FindError, message: FindError.startpath_not_match_find_settings
          end
        end
      File.regular?(p) ->
        if not finder.settings.follow_symlinks and FileUtil.symlink?(p) do
          raise FindError, message: FindError.startpath_not_match_find_settings
        end
        if finder.settings.min_depth > 0 do
          []
        else
          case filter_to_file_results(finder, p) do
            [] -> raise FindError, message: FindError.startpath_not_match_find_settings
            file_results -> file_results
          end
        end
      true -> raise FindError, message: FindError.startpath_not_match_find_settings
    end
  end

  def find_path(finder, path) do
    try do
      path_results = find_path!(finder, path)
      {:ok, path_results}
    rescue
      e in FindError -> {:error, e.message}
    end
  end

  # This is the sequential version of find_files
  defp find_files(finder) do
    # IO.puts("find_files()")
    try do
      results = Enum.map(finder.settings.paths, fn path -> find_path!(finder, path) end)
                |> List.flatten()
      {:ok, FileResultSorter.sort(finder.settings, results)}
    rescue
      e in FindError -> {:error, e.message}
    end
  end

  # This is the asynchronous version of find_files
  defp find_files_async(finder) do
    # IO.puts("find_files_async()")
    try do
      tasks = Enum.map(finder.settings.paths, fn p -> Task.async(fn -> send(self(), find_path!(finder, p)) end) end)
      # # IO.puts("tasks: #{inspect(tasks)})")
      results = Task.await_many(tasks, 1000) |> List.flatten()
      {:ok, FileResultSorter.sort(finder.settings, results)}
    rescue
      e in FindError -> {:error, e.message}
    end
  end

  def find(finder) do
    case validate_settings(finder) do
      {:error, message} -> {:error, message}
      {:ok, _} -> if Enum.count(finder.settings.paths) > 1, do: find_files_async(finder), else: find_files(finder)
    end
  end

  def find!(finder) do
    case find(finder) do
      {:error, message} -> raise FindError, message: message
      {:ok, results} -> results
    end
  end

  # defp readable?(path) do
  #   case File.stat(path) do
  #     {:ok, stat} -> stat.mode && 0o400 != 0
  #     _ -> false
  #   end
  # end

  def validate_settings(finder) do
    if finder.settings.paths == nil or Enum.empty?(finder.settings.paths) do
      {:error, FindError.startpath_not_defined}
    else
      expanded_paths = Enum.map(finder.settings.paths, fn p -> FileUtil.expand_path(p) end)
      cond do
        Enum.any?(expanded_paths, fn p -> !File.exists?(p) end) ->
          {:error, FindError.startpath_not_found}
        # Enum.any?(expanded_paths, fn p -> !readable?(p) end) ->
        #   {:error, "Startpath not readable"}
        not finder.settings.follow_symlinks && Enum.any?(expanded_paths, fn p -> FileUtil.symlink?(p) end) ->
          {:error, FindError.startpath_not_match_find_settings}
        Enum.any?(expanded_paths, fn p -> File.dir?(p) && !traversable_dir?(finder, p) end) ->
          {:error, FindError.startpath_not_match_find_settings}
        Enum.any?(expanded_paths, fn p -> !File.dir?(p) && filter_to_file_results(finder, p) == [] end) ->
          {:error, FindError.startpath_not_match_find_settings}
        finder.settings.min_depth > -1 and finder.settings.max_depth > -1
        and finder.settings.min_depth > finder.settings.max_depth ->
          {:error, FindError.invalid_range_mindepth_and_maxdepth}
        finder.settings.min_size > 0 and finder.settings.max_size > 0
        and finder.settings.min_size > finder.settings.max_size ->
          {:error, FindError.invalid_range_minsize_and_maxsize}
        finder.settings.min_last_mod != nil and finder.settings.max_last_mod != nil
        and DateTime.after?(finder.settings.min_last_mod, finder.settings.max_last_mod) ->
          {:error, FindError.invalid_range_minlastmod_and_maxlastmod}
        true -> {:ok, "Settings are valid"}
      end
    end
  end

  def print_dirs(results, formatter) do
    dirs = Enum.map(results, fn r -> r.path end) |> Enum.uniq() |> Enum.sort()
    if dirs == [] do
      Logging.log("\nMatching directories: 0")
    else
      Logging.log("\nMatching directories (#{Enum.count(dirs)}):")
      formatted_dirs = Enum.map(dirs, fn d -> FileResultFormatter.format_path(formatter, d) end)
      Logging.log("#{Enum.join(formatted_dirs, "\n")}")
    end
  end

  def print_files(results, formatter) do
    if results == [] do
      Logging.log("\nMatching files: 0")
    else
      Logging.log("\nMatching files (#{Enum.count(results)}):")
      formatted_files = Enum.map(results, fn r -> FileResultFormatter.format_file_result(formatter, r) end)
      Logging.log("#{Enum.join(formatted_files, "\n")}")
    end
  end
end
